modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/utilities/constants/strings.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+const StringConstants = {
+	AMPERSAND: '&',
+	AMPERSAND_ENCODED: '&amp;',
+	APOSTROPHE: "'",
+	APOSTROPHE_ENCODED: '&apos;',
+	AT: '@',
+	BACK_SLASH: '\\',
+	BETWEEN: 'BETWEEN',
+	BLANK: '',
+	CARET: '^',
+	CDATA_CLOSE: ']]>',
+	CDATA_OPEN: '<![CDATA[',
+	CLOSE_BRACKET: ']',
+	CLOSE_CURLY_BRACE: '}',
+	CLOSE_PARENTHESIS: ')',
+	COLON: ':',
+	COMMA: ',',
+	COMMA_AND_SPACE: ', ',
+	CONTENT: 'content',
+	DASH: '-',
+	DOLLAR: '$',
+	DOLLAR_AND_OPEN_CURLY_BRACE: '${',
+	DOUBLE_APOSTROPHE: "''",
+	DOUBLE_BACK_SLASH: '\\\\',
+	DOUBLE_CLOSE_BRACKET: ']]',
+	DOUBLE_CLOSE_CURLY_BRACE: '}}',
+	DOUBLE_DASH: '--',
+	DOUBLE_DOLLAR: '$$',
+	DOUBLE_OPEN_BRACKET: '[[',
+	DOUBLE_OPEN_CURLY_BRACE: '{{',
+	DOUBLE_PERIOD: '..',
+	DOUBLE_QUOTE: '""',
+	DOUBLE_SLASH: '//',
+	DOUBLE_SPACE: '  ',
+	DOUBLE_UNDERLINE: '__',
+	EIGHT_STARS: '********',
+	EQUAL: '=',
+	EXCLAMATION: '!',
+	FALSE: 'false',
+	FORWARD_SLASH: '/',
+	FOUR_SPACES: '    ',
+	GRAVE_ACCENT: '`',
+	GREATER_THAN: '>',
+	GREATER_THAN_OR_EQUAL: '>=',
+	INVERTED_EXCLAMATION: '\u00A1',
+	INVERTED_QUESTION: '\u00BF',
+	IS_NOT_NULL: 'IS NOT NULL',
+	IS_NULL: 'IS NULL',
+	ISO_8859_1: 'ISO-8859-1',
+	LAQUO: '&laquo;',
+	LAQUO_CHAR: '\u00AB',
+	LESS_THAN: '<',
+	LESS_THAN_OR_EQUAL: '<=',
+	LIKE: 'LIKE',
+	MINUS: '-',
+	NBSP: '&nbsp;',
+	NEW_LINE: '\n',
+	NOT_EQUAL: '!=',
+	NOT_LIKE: 'NOT LIKE',
+	NULL: 'null',
+	NULL_CHAR: '\u0000',
+	OPEN_BRACKET: '[',
+	OPEN_CURLY_BRACE: '{',
+	OPEN_PARENTHESIS: '(',
+	PERCENT: '%',
+	PERIOD: '.',
+	PIPE: '|',
+	PLUS: '+',
+	POUND: '#',
+	PRIME: '`',
+	QUESTION: '?',
+	QUOTE: '"',
+	QUOTE_ENCODED: '&quot;',
+	RAQUO: '&raquo;',
+	RAQUO_CHAR: '\u00BB',
+	RETURN: '\r',
+	RETURN_NEW_LINE: '\r\n',
+	SEMICOLON: ';',
+	SPACE: ' ',
+	STAR: '*',
+	TAB: '\t',
+	THREE_SPACES: '   ',
+	TILDE: '~',
+	TRIPLE_PERIOD: '...',
+	TRUE: 'true',
+	UNDERLINE: '_',
+	UTC: 'UTC',
+	UTF8: 'UTF-8',
+};
```

Comment:
Forgot to mention. Almost everything in here is currently unused.

Reply:
@wincent 
*TL;DR* I'll remove this.

Yet if you tell me we have a StringPool in JS, we may want to make use of that if needed.

I'll answer this one only Plain copy-paste of our Java StringPool with some neat-up. I went for the simple idea that we have quite a bit of places in Commerce where we compose strings with, say, white spaces, and other things... and and it may not be memory pressuring, but it's nicer to have a single address in memory to refer to than one and the same string/char re-parsed dozens of times. :D

Reply:
> but it's nicer to have a single address in memory to refer to than one and the same string/char re-parsed dozens of times

That sounds very definitely wrong to me.

If you can be bothered, you could benchmark to try and prove a difference, but if you want to save the effort I can tell you that _you won't find any significant difference_. Any difference you find will be buried inside the margins of statistical noise, but note that _even_ if you could measure it, you'd probably get the exact opposite of the result you're expecting; if you "save" the parsing time of re-parsing something like `'/'` everywhere its used, remember that you'll _instead_ be spending time re-parsing constant names like `FORWARD_SLASH`, and worse than that, invoking the module machinery to access the `StringPool` object and then doing a property look up for every access. It's actually doing _more_ work, not less.

That means that _performance_ shouldn't be the dimension you use to add a pattern like this (and if you really did care about it, you wouldn't ship dozens of strings that you aren't even going to reference 😉 ); it should be _readability_ and _writeability_. So, on the readability/writeability axis, things like `BLANK` are very obviously worse (due to ambiguity), things like `FORWARD_SLASH` are going to contribute towards needless wrapping, and things like `INVERTED_QUESTION` are a win because nobody wants to have to remember how to write that. ie. in here you've got the good, the bad, and the ugly.

Reply:
Just to understand though, set aside the non-existent significant execution time differences... I may get that the property look-up in something like an exported (and re-exported)`StringPool` object may not be ideal, especially 'cause 1% of it was used. But still each and all of those keys could actually be exported with a simple `export const CONSTANT = '...';` and at that point I'd see my argument slightly more sound, given tree-shaking and obfuscation (the `CONSTANT` may become a simple `a`) which should take place at build-time (hence making the parser's life easier at exec time -- though insignificantly, I get it).

Correct or...? Just out of discussion.

Reply:
I like the argument from tree-shaking (even though it is still only a theoretical benefit for us at this point because we don't/can't really do it properly due to the way we dynamically bundle resources — ie. due to dynamism of portlets, we can't do ahead-of-time tree-shaking, and we're stuck with module-level granularity for anything we do at runtime in the current system).

But _even if_ your compiler is smart enough to turn `'/'` into an `a`, that `a` is still coming to you via a module import which is not free. You're paying a relatively large amount of overhead just for the pleasure of doing `import {anything} from 'anything-else'`. As an example of the current architecture, look at all the stuff we do just to set-up an AMD module and provide it with access to its dependencies (and _whatever_ we do in the future will have _some_ residual amount of boilerplate-ish overhead just to wire things up):

```js
Liferay.Loader.define("layout-content-page-editor-web@3.0.0/page_editor/core/createDNDBackend", ['module', 'exports', 'require', 'frontend-js-react-web$react-dnd-html5-backend'], function (module, exports, require) {
  var define = undefined;
  var global = window;
  {
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.default = createDNDBackend;

    // (omitted... the code you actually care about)
  }
});
```

(Just a module selected at random from liferay-portal which imports one thing and exports one thing.)

So, I don't think performance is the right angle to be looking at this from for these tiny strings. It's still readability/writability like I said earlier, and when you look at it through that lens, it's pretty easy to make a case that the indirection of things like `StringPool.BLANK` is harmful, in contrast to counter-examples like `RAQUO_CHAR` and `INVERTED_EXCLAMATION`, which are much easier to defend.

Reply:
Crystal clear and thanks for your time to reply. Yeah I was kind of expecting that the 2 possible counter-arguments would indeed be: "that's Liferay" and (independently from Liferay) "AMD doesn't come for free". Modules are still objects at the end of the day and not having tree-shaking at our disposal (at the moment?), for sure is painful. 

I decidedly didn't go on to defend the readability part because: 1) I really wanted more substantial argumentation related to performance, our infra and the inherent why's. :p 2) I was fully agreeing with you ever since the very beginning, being a bare copy-paste of the StringPool Java class that I wasn't even supposed to commit in full.

So, that's why I thank you twice for taking the time to bear with me on the topic. I owe you a 🍺 .

Reply:
> being a bare copy-paste of the StringPool Java class

Well, TIL... this lead me to read the Java implementation today which I'd only ever seen cursorily before, and I was horrified to see how all those names that I criticized are actually from there 😂. Likewise, I looked in the related `CharPool` class, and will file that away for another day if it comes up. 😁